src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorPreCompiler.cs

```diff
@@ -58,93 +64,126 @@ public virtual void CompileViews([NotNull] IBeforeCompileContext context)
+            var filesToProcess = new List<RelativeFileInfo>();
+            GetFileInfosRecursive(root: string.Empty, razorFiles: filesToProcess);
 
-            foreach (var info in GetFileInfosRecursive(string.Empty))
-            {
-                var descriptor = ParseView(info, context);
+            var razorFiles = new RazorFileInfo[filesToProcess.Count];
+            var syntaxTrees = new SyntaxTree[filesToProcess.Count];
+            var parallelOptions = new ParallelOptions { MaxDegreeOfParallelism = MaxDegreesOfParallelism };
 
-                if (descriptor != null)
+            Parallel.For(0, filesToProcess.Count, parallelOptions, index =>
+            {
+                var file = filesToProcess[index];
+                var cacheEntry = PreCompilationCache.GetOrSet(file.RelativePath,
+                                                              file,
+                                                              OnCacheMiss);
+                if (cacheEntry != null)
                 {
-                    list.Add(descriptor);
+                    if (cacheEntry.Success)
+                    {
+                        syntaxTrees[index] = cacheEntry.SyntaxTree;
+                        razorFiles[index] = cacheEntry.FileInfo;
+                    }
+                    else
+                    {
+                        lock (context.Diagnostics)
+                        {
+                            foreach (var diagnostic in cacheEntry.Diagnostics)
+                            {
+                                context.Diagnostics.Add(diagnostic);
+                            }
+                        }
+                    }
                 }
-            }
+            });
+
+            context.CSharpCompilation = context.CSharpCompilation
+                                               .AddSyntaxTrees(syntaxTrees.Where(tree => tree != null));
+            return razorFiles.Where(file => file != null);
+        }
 
-            return list;
+        protected IMvcRazorHost GetRazorHost()
+        {
+            return _serviceProvider.GetRequiredService<IMvcRazorHost>();
         }
 
-        private IEnumerable<RelativeFileInfo> GetFileInfosRecursive(string currentPath)
+        private PrecompilationCacheEntry OnCacheMiss(ICacheSetContext cacheSetContext)
         {
-            var path = currentPath;
+            var fileInfo = (RelativeFileInfo)cacheSetContext.State;
+            var entry = GetCacheEntry(fileInfo);
 
-            var fileInfos = _fileSystem.GetDirectoryContents(path);
-            if (!fileInfos.Exists)
+            if (entry != null)
             {
-                yield break;
+                cacheSetContext.AddExpirationTrigger(_fileSystem.Watch(fileInfo.RelativePath));
+                foreach (var viewStartPath in ViewStartUtility.GetViewStartLocations(fileInfo.RelativePath))
+                {
+                    cacheSetContext.AddExpirationTrigger(_fileSystem.Watch(viewStartPath));
```

Comment:
Is there any consideration for compilation settings change, or changing project.json just invalidates everything?